### ISSUE TITLE
Generate URL automatically based on config

### ISSUE THREAD

**Issue by jacobemcken** [jacobemcken/aws-simple-sign#4]
Currently, the user of the library is responsible for creating the URL for the S3 object.

Advantages: Allows the user to choose between (the old) path-style URLs vs. (the new) virtual hosted-style URLs:
See: https://aws.amazon.com/blogs/aws/amazon-s3-path-deprecation-plan-the-rest-of-the-story/
This is especially nice for bucket names with dots (.) in its name, causing issues with TLS and SSL certificates.

Disadvantages: The user has to construct URL's that the library could just as well create since it is reading `AWS_*` environment variables and config from `.aws/*` anyway.

**Comment by jacobemcken** [comment]
Some initial work:
https://github.com/jacobemcken/aws-simple-sign/pull/5

Worth testing with MinIO:
https://stackoverflow.com/questions/59693471/how-to-setup-minio-server-to-use-virtual-hosted-style

**Comment by christoffer-ludvigsen** [comment]
i could not get access to push to any branch so here is the code

```clojure
(defn presign
  "Take an URL for a S3 object and returns a string with a presigned GET-URL
   for that particular object.
   Takes the following options (a map) as the last argument,
   the map value shows the default values:
       {:ref-time (java.util.Date.) ; timestamp incorporated into the signature
        :expires \"3600\"           ; signature expires x seconds after ref-time
        :region \"us-east-1\"}      ; signature locked to AWS region
   
   By default credentials are read from standard AWS location."
  ([url]
   (presign url (read-env-credentials) {}))
  ([url opts]
   (presign url (read-env-credentials) opts))
  ([url credentials {:keys [ref-time region expires]
                     :or {expires "3600"
                          region (:aws/region credentials)
                          ref-time (Date.)}}]
   (let [url-obj (URL. url)
         host (.getHost url-obj)
         port (.getPort url-obj)
         host-port (if (pos? port)
                     (str host ":" port)
                     host)
         service "s3"
         timestamp (.format formatter (.toInstant ^Date ref-time))
         scope (str (subs timestamp 0 8) "/" region "/" service "/aws4_request")
         query-params (conj {"X-Amz-Algorithm" algorithm
                             "X-Amz-Credential" (str (:aws/access-key credentials) "/" scope)
                             "X-Amz-Date" timestamp
                             "X-Amz-SignedHeaders" "host"}
                            (when-let [token (:aws/token credentials)]
                              ["X-Amz-Security-Token" token])
                            (when expires
                              ["X-Amz-Expires" expires]))
         signature (signature (.getPath url-obj)
                              credentials
                              {:timestamp timestamp
                               :region region
                               :service service
                               :scope scope
                               :query-params query-params
                               :signed-headers {"host" host-port}})]
     (str (.getProtocol url-obj) "://" host-port (.getPath url-obj) "?"
          (->query-str query-params)
          "&X-Amz-Signature=" signature))))

; TODO figure out what to do with alternative ports like MinIO: localhost:9000
(defn generate-presigned-url
  "Takes bucket name, object key and a map with options.
   The option `:path-style true` forces the use of the 'old way' of URL's."
  ([bucket object-key opts]
   (generate-presigned-url bucket object-key (read-env-credentials) opts))
  ([bucket object-key credentials
    {:keys [region endpoint]
     :or {region (:aws/region credentials)
          endpoint (:aws/endpoint credentials)} :as opts}]
   (let [endpoint' (or endpoint ; make sure it ends with slash (/)
                       (str "https://s3." region ".amazonaws.com/"))]
     (-> (if (:path-style opts)
           (str endpoint' bucket "/")
           (str/replace endpoint' #"://" (str "://" bucket ".")))
         (str object-key)
         (presign credentials opts)))))
```

**Comment by jacobemcken** [comment]
Resolved with: #5 

Thanks for the help @christoffer-ludvigsen
